Expand the got state changing functionality

zmyrgel opened the pull request:
Implement some missing commands and refactors the got-merge functionality.

zmyrgel committed:
add one not implemented function to status list

zmyrgel committed:
remove duplicate vc-got-create-repo function

zmyrgel committed:
implement vc-got-checkin-patch command

zmyrgel committed:
implement vc-got-revert-files

This is a bit of performance improvement, we can revert operations on
multiple files with one command instead of reverting single files in a loop.

zmyrgel committed:
implement vc-got-get-change-comment

zmyrgel committed:
drop comments from status list

its easier to compare it with other status lists.

zmyrgel committed:
implement working-tree functionality

vc has added working-tree functionality to work with git-worktree(1).
Got does not use exact functionality but these map quite nicely to Got
work trees.

- vc-got-add-working-tree creates a new work tree with "got checkout"

- vc-got-delete-working-tree removes uuid from got repository and
  deletes the directory

- vc-got-move-working-tree just moves the directory to new place.

The git-worktree also tracks the locations of the working trees and
there is vc-known-other-working-trees command to list them. Got does not
track checkout locations so this is hard to implement currently.
When creating working-tree from vc, we could add our own file to got
repository to track these work tree locations to got some tracking to
these directories but this is left for later.

zmyrgel committed:
refactor merging

make vc-got-merge run the actual "got merge" command and track the merge
status.
The got has more commonly used "got integrate" command so add a separate
command for it.

omar-polo commented on vc-got.el:
```diff
@@ -637,6 +638,26 @@ populates it with files from a directory polled from user."
+            ;; heuristic to figure out what files are affected by the patch
+            (goto-char (point-min))
+            (while (re-search-forward "^file \+" nil t)
+              (forward-char 2)
```
nit: can't you add a space at the end of the `re-search-forward` so that we can skip the forward-char?

zmyrgel replied:
Yeah, sounds better.
But while testing this I noticed that "file +" don't always seem to have the file listed. 
For example when deleting file it looks like this:
```
commit - 6fe3bffb0a2bcd31c562235abd0f864d628686d6
blob - 74f32b27f3823546f998d8cea8ca7334dbd4e4a0
file + /dev/null
--- editors/emacs/pkg/DESCR
+++ /dev/null
@@ -1,27 +0,0 @@
```
Should we try to match the "^---" but that won't work correctly for exampl yaml files which use --- as record separator. Do we need better parser here?

zmyrgel replied:
Would this look better here?

```
 vc-got.el | 19 +++++++++++++------
 1 file changed, 13 insertions(+), 6 deletions(-)

diff --git a/vc-got.el b/vc-got.el
index b2c4e75..78be4e4 100755
--- a/vc-got.el
+++ b/vc-got.el
@@ -651,6 +651,14 @@ populates it with files from a directory polled from user."
                    '(("Author" . "-A"))
                    comment)))
 
+(defconst vc-got--commit-header-re
+  (concat "^commit - \\([a-z0-9]+\\)\n"
+          "blob - \\([a-z0-9]+\\)\n"
+          "file \\+ .+\n"
+          "--- \\(.+\\)\n"
+          "\\+\\+\\+ \\(.+\\)\n")
+  "Regexp to match commit headers.")
+
 (defun vc-got-checkin-patch (patch-string comment)
   "Commit PATCH-STRING with COMMENT to repository."
   (let ((temp-patch-file (make-temp-file "got-patch")))
@@ -660,14 +668,13 @@ populates it with files from a directory polled from user."
             (insert patch-string)
             ;; heuristic to figure out what files are affected by the patch
             (goto-char (point-min))
-            (while (re-search-forward "^file \+" nil t)
-              (forward-char 2)
-              (push (buffer-substring-no-properties (point) (line-end-position))
-                    files)))
+            (while (re-search-forward vc-got--commit-header-re nil t)
+              (push (match-string-no-properties 3) files)))
           ;; apply the patch to work tree
           (vc-got-command nil 0 temp-patch-file "patch")
-          ;; commit the changed files
-          (vc-got-checkin files comment))
+          ;; commit the changed files, reverse to keep same order as in
+          ;; patch file
+          (vc-got-checkin (reverse files) comment))
       ;; delete the temp patch, no longer needed
       (delete-file temp-patch-file))))
 
```

zmyrgel replied:
And a another thing, should we have more validation here to ensure that the patch-string really contains a patch? Or is it ensured by the "upper" layers that proper patch is used when calling this.

omar-polo replied:
> Should we try to match the "^---" but that won't work correctly for exampl yaml files which use --- as record separator. Do we need better parser here?

shouldn't be a problem, as yaml file in diffs would have their `---` be prefixed with a space, a plus, or another minus.

> Would this look better here?

yeah, thanks, looks way better to me!

> And a another thing, should we have more validation here to ensure that the patch-string really contains a patch? Or is it ensured by the "upper" layers that proper patch is used when calling this.

not sure, the contract seems to imply that we receive a patch, so I don't think I would bother.

omar-polo commented on vc-got.el:
```diff
@@ -753,6 +754,19 @@ It's like `vc-process-filter' but supports \\r inside S."
   "Execute a send prompting for the full command if PROMPT is not nil."
   (vc-got--push-pull vc-got-program "send" prompt))
 
+(defun vc-got-get-change-comment (_files rev)
+  "Return the change comments given REV.  The files argument is ignored."
+  (with-temp-buffer
+    (vc-got--log nil 1 rev)
+    (goto-char (point-min))
+    (forward-line 4)
+    (save-excursion
+      (while (not (eobp))
+        (delete-char 1)
+        (forward-line))
+      (delete-blank-lines))
+    (buffer-substring-no-properties (point) (point-max))))
```
can we add some comments here?  I was a bit confused while reading it, took some time to understand.

To be clear, I think it's correct, but a comment on `(forward-line 4)` with something like "skip commit header" and annotate the save-excursion loop with "delete the one-space indentation" would help =)

side question: i'm still unsure about the `delete-blank-lines`.

omar-polo commented on vc-got.el:
```diff
@@ -763,6 +766,40 @@ It's like `vc-process-filter' but supports \\r inside S."
+      (insert-file-contents (expand-file-name ".got/uuid" root-dir))
+      (string-trim (buffer-string)))))
+
+(defun vc-got-add-working-tree (dir)
+  "Checkout a new work tree to DIR."
+  ;; NOTE: `got' tracks the work trees but not their locations, should
+  ;; we re-use the repo .got/refs/got/worktree/ directory to store the
+  ;; checkout directory? This would allow to implement
+  ;; `known-other-working-trees' to list known work trees and jump
+  ;; between them.
+  (when-let* ((branch (vc-got--prompt-branch "Create new work tree from branch: ")))
+    (vc-got-command nil 0 dir "checkout" "-b" branch)))
```
random comment while reviewing: this would be interesting as a feature but seems brittle.  if a user moves a worktree we have to provide ways to update this metadata etc..

zmyrgel replied:
Yeah, its brittle but it would give something to work with. We could track all work trees generated by emacs vc but everything else would be "hidden". Though we could add extra command "vc-got-mark-as-work-tree" which would check that current directory is a got work tree and it would ensure the got repository would have up-to-date reference to it. For cases when the directory is created outside emacs but user wishes to use it from emacs.
 
I'm not sure how well git-worktree tracks the directories. Got seems to track the checkouts somewhat by generating uuid for those. I'm not sure how/when those are cleared from the repository.

omar-polo replied:
The refs need to be removed manually like you're doing with got.  usually it's not a problem if you forget to do so, since one usually ends up with just a handful of worktrees, and the wasted space for the unused refs is very small.

Other than that I don't mind the idea of tracking the paths, I just feel a bit unsure about going to add files to the git repository that can be mistaken by symrefs by tools (since they'd be under the refs/ namespace, and in particular under the `refs/got` which is "owned" by the tool)

omar-polo commented on vc-got.el:
```diff
@@ -763,6 +766,40 @@ It's like `vc-process-filter' but supports \\r inside S."
+  "Delete a work tree in directory DIR."
+  (if-let* ((uuid (vc-got--work-tree-uuid dir)))
+      (with-temp-buffer
+        (vc-got-command t 0 (concat "refs/got/worktree/base-" uuid) "ref" "-d")
```
is the order wrong here?  shouldn't it be `... "ref" "-d" (concat ...)`?

zmyrgel replied:
The "(concat ...)" part is the file argument so it is placed last in the command. So it should be correct.

omar-polo replied:
right!  I misread `vc-do-command`.

omar-polo commented on vc-got.el:
```diff
@@ -763,6 +766,40 @@ It's like `vc-process-filter' but supports \\r inside S."
+  (if (file-exists-p (expand-file-name ".got" from))
+      (progn
+        (copy-directory from to)
+        (delete-directory from))
```
not sure what to think about this copy+delete thing.  don't we have a `mv` available?  (we can actually rely on that, as got does not work on windows anyway)

zmyrgel replied:
Seems that rename-file should be able to rename directories as well so we could use that here.

omar-polo replied:
thanks!

omar-polo commented on vc-got.el:
```diff
@@ -745,6 +773,7 @@ It's like `vc-process-filter' but supports \\r inside S."
 ;; TODO: this could be expanded.  After a pull the worktree needs to
 ;; be updated, either with a ``got update -b branch-name'' or ``got
 ;; update -b remote/branchname'' plus a rebase.
+;; TODO: can this handle prefix argument? pull does fetch, C-u pull does update?
```
I guess that here we should do some rebase or merging, possibly letting the user choose with an option or prefix argument, but don't use the prefix to do the update, that should be done by default I think.

omar-polo reviewed (commented):
sorry for the delay, here's a first round of review on this.

zmyrgel committed:
use rename-file instead of copy-directory+delete-directory

zmyrgel committed:
add few comments to vc-got-get-change-comment

zmyrgel committed:
improve vc-got-checkin-patch file heuristic

the "^file" lines do not seem to give out the file when a patch is
deleting something. The '---' header seems to give the file for all
cases so use it to determine the affected file.

Add vc-got--commit-header-re to match the header block and pick the
wanted information out of it.

omar-polo reviewed (approved):
looks good to me, looking forward to use it day to day =)

let me know if you want to tweak it more or if I can merge it.

zmyrgel commented:
Yeah, I think this can be merged as it is. These can be extended and improved later if needed.

omar-polo commented:
Thanks!
